### PR TITLE
feat: enhance `Update-NsxtManagerPasswordComplexity` for VCF 5.0 and NSX 4.x

### LINKED ISSUE #38: `Update-NsxtManagerPasswordComplexity` cmdlet does not work for NSX 4.x
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

`Update-NsxtManagerPasswordComplexity` cmdlet was not updating the configuration for the corresponding passwordpolicy.
Upon checking, it is learnt that on VCF5.0, direct update of common-password file is not supported instead configuration is to be pushed only using NSX CLI's.
This requires Document change as well.

### Use Case(s)

```powershell
Execute the following cmdlets and check if configurations are changed.

Update-NsxtManagerPasswordComplexity -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass -domain $sddcDomainName -minLength $minLength -minLowercase $minLowercase -minUppercase $minUppercase -minNumerical $minNumerical -minSpecial $minSpecial -minUnique $minUnique -maxRetry $maxRetry

Request-NsxtManagerPasswordComplexity   -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass   -domain $sddcDomainName

```

### Potential Configuration

Now the NSX-T API "**https://\<NSX-T manager\>/api/v1/node/aaa/auth-policy**" itself can be used to read the parameters related to password complexity policy as well. So extending the use of this API, will suffice the need.

### References

_No response_

### PR DESCRIPTION
**Pull request by sowjuec**
**Summary of Pull Request**

Enhanced `Update-NsxtManagerPasswordComplexity` to handle VCF 5.0 and NSX 4.x changes.

**Type of Pull Request**

- [ ] This is a bug fix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is a chore update
- [ ] This is something else.
      Please describe:

**Related to Existing Issues**

Closes #38

**Test and Documentation Coverage**

- [x] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 1]
```diff
@@ -138,7 +138,7 @@ Function Invoke-PasswordPolicyManager {
         if (Test-VCFConnection -server $sddcManagerFqdn) {
             if (Test-VCFAuthentication -server $sddcManagerFqdn -user $sddcManagerUser -pass $sddcManagerPass) {
                 if (!(Test-Path -Path $reportPath)) {Write-Warning "Unable to locate report path $reportPath, enter a valid path and try again"; Write-Host ""; Break }
-                if (!(Test-Path -Path $($reportPath + '\' + $policyFile))) {Write-Warning "Unable to locate policy file $policyFile, enter a valid path and try again"; Write-Host ""; Break }
+                if (!(Test-Path -Path $($reportPath + '\' + $policyFile))) {Write-Warning "Unable to locate policy file $policyFile, enter a valid path and try again"; Write-Host ""; Break }
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:141]: Looks like a unneeded whitespace was added here.

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 2]
```diff
@@ -592,9 +599,14 @@ Function Get-PasswordPolicyDefault {
 
     Param (
         [Parameter (Mandatory = $false, ParameterSetName = 'json')] [ValidateNotNullOrEmpty()] [Switch]$generateJson,
+        [Parameter (Mandatory = $true)] [ValidateSet('4.4.0','4.5.1','5.0.0')] [String]$version='5.0.0',
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:602]: Are we sure we want to default to 5.0.0 or require the version to be set?

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 3]
```diff
@@ -793,6 +816,7 @@ Export-ModuleMember -Function Get-PasswordPolicyDefault
 Function Get-PasswordPolicyConfig {
     Param (
         [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$reportPath,
+        [Parameter (Mandatory = $true)] [ValidateSet('4.4.0','4.5.1','5.0.0')] [String]$version='5.0.0',
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:819]: Are we sure we want to default to 5.0.0 or require the version to be set?

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 4]
```diff
@@ -801,19 +825,24 @@ Function Get-PasswordPolicyConfig {
...
+            if ($customConfig.vcf.vcfVersion -eq $version) {
+                $result = Test-PasswordPolicyConfig -customConfig $customConfig -version $version
+                if ($result -eq "true") {
+                    Write-Output "Validation of Password Policy Configuration File: PASSED"
+                } else {
+                    Write-Error "Validation of Password Policy Configuration File: FAILED"
+                    Break
+                }
+            } else {
+                Write-Error "Password Policy Configuration File Version is $($customConfig.vcf.vcfVersion) and version passed is $version : FAILED "
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:837]:
```suggestion
                Write-Error "Password Policy Configuration File Version is $($customConfig.vcf.vcfVersion) and version provided is $version: FAILED "
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 5]
```diff
@@ -2326,17 +2383,21 @@ Function Request-SsoPasswordComplexity {
...
-        }
-    }
 
 	Try {
 		if (Test-Connection -server $server) {
-			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {              
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:2392]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 6]
```diff
@@ -2444,17 +2507,22 @@ Function Request-SsoAccountLockout {
...
-        }
-    }
+
 
 	Try {
 		if (Test-Connection -server $server) {
 			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:2517]:
```suggestion
                    if(((Get-VCFManager) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 7]
```diff
@@ -2994,17 +3070,20 @@ Function Request-VcenterPasswordExpiration {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).vcenterServer.passwordExpiration
-        }
-    }
-
 	Try {
         if (Test-VCFConnection -server $server) {
             if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:3078]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 8]
```diff
@@ -3113,10 +3196,14 @@ Function Request-VcenterPasswordComplexity {
                         if (Test-vSphereConnection -server $($vcfVcenterDetails.fqdn)) {
                             if (Test-vSphereAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
                                 if ($drift) {
+                                    $version = ""
+                                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:3200]:
```suggestion
                                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 9]
```diff
@@ -3601,17 +3687,20 @@ Function Request-VcenterRootPasswordExpiration {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).vcenterServerLocal.passwordExpiration
-        }
-    }
-
 	Try {
         if (Test-VCFConnection -server $server) {
             if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:3695]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 10]
```diff
@@ -4295,17 +4391,22 @@ Function Request-NsxtManagerPasswordComplexity {
...
-    }
+
 
 	Try {
         if (Test-VCFConnection -server $server) {
-            if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+            if (Test-VCFAuthentication -server $server -user $user -pass $pass) {                
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:4401]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 11]
```diff
@@ -4406,17 +4531,20 @@ Function Request-NsxtManagerAccountLockout {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).nsxManager.accountLockout
-        }
-    }
-
     Try {
         if (Test-VCFConnection -server $server) {
             if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:4539]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 12]
```diff
@@ -5145,17 +5350,20 @@ Function Request-NsxtEdgePasswordExpiration {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).nsxManager.passwordExpiration
-        }
-    }
-
 	Try {
         if (Test-VCFConnection -server $server) {
             if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:5358]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 13]
```diff
@@ -5357,17 +5571,22 @@ Function Request-NsxtEdgeAccountLockout {
...
-        }
-    }
+
 
     Try {
         if (Test-VCFConnection -server $server) {
             if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:5581]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 14]
```diff
@@ -6091,17 +6313,20 @@ Function Request-EsxiPasswordExpiration {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).esxi.passwordExpiration
-        }
-    }
-
 	Try {
 		if (Test-Connection -server $server) {
 			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:6321]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 15]
```diff
@@ -6210,17 +6438,20 @@ Function Request-EsxiPasswordComplexity {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).esxi.passwordComplexity
-        }
-    }
-
 	Try {
 		if (Test-Connection -server $server) {
 			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:6446]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 16]
```diff
@@ -6333,17 +6567,20 @@ Function Request-EsxiAccountLockout {
...
-            $requiredConfig = (Get-PasswordPolicyConfig).esxi.accountLockout
-        }
-    }
-
 	Try {
 		if (Test-Connection -server $server) {
 			if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:6575]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 17]
```diff
@@ -7206,14 +7460,13 @@ Function Request-WsaLocalUserAccountLockout {
...
                         }
-                        Disconnect-VIServer $vcfVcenterDetails.fqdn -Confirm:$false -WarningAction SilentlyContinue
                     }
                 }
             }
         }
 	} Catch {
         Debug-ExceptionWriter -object $_
-    }
+    }
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:7469]:
```suggestion
    }
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 18]
```diff
@@ -7988,18 +8243,21 @@ Function Request-LocalUserPasswordExpiration {
...
-        }
-        $requiredConfig = Invoke-Expression $command
-    }
-
 	Try {
         if (Test-VCFConnection -server $server) {
             if (Test-VCFAuthentication -server $server -user $user -pass $pass) {
+                if ($drift) {
+                    $version = ""
+                    if(((get-vcfmanager).version) -match "\d+\.\d+\.\d+") {
```
Comment (tenthirtyam) [inline VMware.CloudFoundation.PasswordManagement.psm1:8251]:
```suggestion
                    if(((Get-VCFManager).version) -match "\d+\.\d+\.\d+") {
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Minor changes requested.

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Merge will be dependent on https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/pull/260 which is planned for [`PowerValidatedSolutions` Milestone v2.4.0](https://github.com/vmware-samples/power-validated-solutions-for-cloud-foundation/milestone/11).

This will require an update to `Test-VcfPasswordManagementPrereq` to set:

```powershell
        $modules = @(
            @{ Name=("VMware.PowerCLI"); MinimumVersion=("13.0.0")}
            @{ Name=("VMware.vSphere.SsoAdmin"); MinimumVersion=("1.3.9")}
            @{ Name=("PowerVCF"); MinimumVersion=("2.3.0")}
            @{ Name=("PowerValidatedSolutions"); MinimumVersion=("2.4.0")}
        )
```

File: VMware.CloudFoundation.PasswordManagement.psm1 (hunk 2 above)
Comment (sowjuec) [reply VMware.CloudFoundation.PasswordManagement.psm1:602]: Yes. Gary confirmed default has to be set and has to be 5.0.0

File: VMware.CloudFoundation.PasswordManagement.psm1 (hunk 3 above)
Comment (sowjuec) [reply VMware.CloudFoundation.PasswordManagement.psm1:819]: Yes, I had discussed with Gary and he insisted, it has to be mandatory parameter with default value being 5.0.0 in place.

**Comment by tenthirtyam** [comment]
Hi @sowjuec - could you make each conversation as resolved?

It seems that the change on line 837 broke the CI test.

```bash
Error: [-] /home/runner/work/powershell-module-for-vmware-cloud-foundation-password-management/powershell-module-for-vmware-cloud-foundation-password-management/.ci/pester.tests.ps1 failed with:

Message

  ActionPreferenceStopException: The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: At /home/runner/work/powershell-module-for-vmware-cloud-foundation-password-management/powershell-module-for-vmware-cloud-foundation-password-management/VMware.CloudFoundation.PasswordManagement.psm1:837 char:132

  + … ustomConfig.vcf.vcfVersion) and version provided is $version: FAILED  …

  +                                                       ~~~~~~~~~

  Variable reference is not valid. ':' was not followed by a valid variable name character. Consider using ${} to delimit the name.

  at <ScriptBlock>, /home/runner/work/powershell-module-for-vmware-cloud-foundation-password-management/powershell-module-for-vmware-cloud-foundation-password-management/.ci/pester.tests.ps1:2

Tests completed in 592ms

Tests Passed: 0, Failed: 2, Skipped: 0 NotRun: 0

Container failed: 1

  - /home/runner/work/powershell-module-for-vmware-cloud-foundation-password-management/powershell-module-for-vmware-cloud-foundation-password-management/.ci/pester.tests.ps1

Error: Process completed with exit code 1.
```

**Pull request merged**